src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlTree.java

```diff
@@ -177,8 +171,8 @@ public HtmlTree add(Content content) {
         if (content instanceof ContentBuilder) {
             ((ContentBuilder) content).contents.forEach(this::add);
         }
-        else if (content == HtmlTree.EMPTY || content.isValid()) {
-            // quietly avoid adding empty or invalid nodes (except EMPTY)
+        else if (content.isValid()) {
```

Comment:
Should the content builder have a similar validity check to ensure if it's not empty, its contents are always valid? otherwise it's quite hard to define if the content builder is valid or not as it can be considered either and always need special case in client code. In comparison, the html tree's contents are always valid no matter if the outer tags are valid or not.

Reply:
I think the current behaviour of `ContentBuilder` to accept all content and check validity on demand is the better solution. At least in theory it's possible that invalid content (e.g. an empty HtmlTree) is added to a ContentBuilder which later becomes valid by adding valid content to it. On the other hand, one could argue that `ContentBuilder` and `HtmlTree` should behave the same, which would speak for checking validity when adding to `ContentBuilder`.

Reply:
Can we take the bigger step and make it unnecessary to have/use `isValid`?  We should never knowingly create nodes that will be invalid by whatever metric we choose to define validity.  If there's a risk of creating nodes that will end up being invalid, we should do the work ahead of time to not create the nodes in the first place.

Bottom line, the original reasons for `isValid` were, at best, flawed. Is this the time to fix that?

Reply:
> Can we take the bigger step and make it unnecessary to have/use `isValid`?

Yes.

> the original reasons for `isValid` were, at best, flawed. Is this the time to fix that?

When you say "time", do you mean "this PR" or "physical time"? If you meant the latter, then I find it hard to imagine a simple and correct fix that would also fit in the JDK 17 timeframe.

Reply:
I agree that our current `isValid` implementation does not feel right. 

But there are many (dozens?) of places in the code that potentially generate empty elements. If we want to avoid these, shifting the responsability to the content producer will blow up the code a lot. I'm trying out different things to see if I can come up with a better solution.